fdbserver/VersionedBTree.actor.cpp

```diff
@@ -2541,11 +2598,44 @@ class DWALPager final : public IPager2 {
 		if (self->memoryOnly) {
 			return Void();
 		}
+
+		// If a truncation up to include pageID has not yet been completed
+		if (pageID >= self->filePageCount) {
+			// And no extension pending will include pageID
+			if (pageID >= self->filePageCountPending) {
+				// Then start an extension that will include pageID
+				self->fileExtension = extendToCover(self, pageID);
+			}
+
+			// Wait for extension that covers pageID to complete;
+			wait(self->fileExtension);
+		}
+
 		// Note:  Not using forwardError here so a write error won't be discovered until commit time.
 		wait(self->pageFile->write(data, blockSize, (int64_t)pageID * blockSize));
 		return Void();
 	}
 
+	ACTOR static Future<Void> extendToCover(DWALPager* self, uint64_t pageID) {
+		// Calculate new page count, round up to nearest multiple of growth size > pageID
+		state int64_t newPageCount = pageID + SERVER_KNOBS->REDWOOD_PAGEFILE_GROWTH_SIZE_PAGES -
+		                             (pageID % SERVER_KNOBS->REDWOOD_PAGEFILE_GROWTH_SIZE_PAGES);
+
+		// Indicate that extension to this new count has been started
+		self->filePageCountPending = newPageCount;
+
+		// Wait for any previous extensions to complete
+		wait(self->fileExtension);
```

Comment:
Just to confirm my understanding, the reason this works along with line 2607 updating self->FileExtension to this future is because extendToCoverwill run up to the wait, including passing the current self->fileExtension to the wait, before updating self->fileExtension on line 2607?
This seems like a weird interaction with futures that might at least warrant a comment or something?